cpp/include/cuspatial/experimental/detail/point_polygon_distance.cuh

```diff
@@ -0,0 +1,170 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/cuda_utils.hpp>
+#include <cuspatial/detail/iterator.hpp>
+#include <cuspatial/detail/utility/device_atomics.cuh>
+#include <cuspatial/detail/utility/linestring.cuh>
+#include <cuspatial/detail/utility/offset_to_keys.cuh>
+#include <cuspatial/detail/utility/zero_data.cuh>
+#include <cuspatial/error.hpp>
+#include <cuspatial/experimental/detail/algorithm/is_point_in_polygon.cuh>
+#include <cuspatial/experimental/ranges/range.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/functional.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/reduce.h>
+#include <thrust/tabulate.h>
+
+#include <cstdint>
+#include <limits>
+#include <type_traits>
+
+namespace cuspatial {
+namespace detail {
+
+/**
+ * @brief For each point in the multipoint, compute point-in-multipolygon in corresponding pair.
+ */
+template <typename MultiPointRange, typename MultiPolygonRange>
+struct point_in_multipolygon_test_functor {
+  MultiPointRange multipoints;
+  MultiPolygonRange multipolygons;
+
+  point_in_multipolygon_test_functor(MultiPointRange multipoints, MultiPolygonRange multipolygons)
+    : multipoints(multipoints), multipolygons(multipolygons)
+  {
+  }
+
+  template <typename IndexType>
+  uint8_t __device__ operator()(IndexType pidx)
+  {
+    auto point        = thrust::raw_reference_cast(multipoints.point(pidx));
+    auto geometry_idx = multipoints.geometry_idx_from_point_idx(pidx);
+
+    bool intersects = false;
+    for (auto polygon : multipolygons[geometry_idx]) {
+      intersects = intersects || is_point_in_polygon(point, polygon);
+    }
+    return static_cast<uint8_t>(intersects);
+  }
+};
+
+/**
+ * @brief Kernel to compute the distance between pairs of point and polygon.
+ */
+template <class MultiPointRange,
+          class MultiPolygonRange,
+          class IntersectionRange,
+          class OutputIterator>
+void __global__ pairwise_point_polygon_distance_kernel(MultiPointRange multipoints,
+                                                       MultiPolygonRange multipolygons,
+                                                       IntersectionRange intersects,
+                                                       OutputIterator distances)
+{
+  using T = typename MultiPointRange::element_t;
+
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x; idx < multipolygons.num_points();
```

Comment:
So iterating over segments in all multipolygons? Can this loop be made clearer with one of your refs?